### PR TITLE
Fix InputEventKey test using incorrect equality logic

### PR DESCRIPTION
**Pull request by ZiedKs**
### Summary
Fixes an incorrect InputEventKey test that passed despite testing the wrong
behavior, due to using instance equality instead of keycode matching.

### Details
- `InputEventKey::is_match()` compares keycodes, not object instances
- The previous test set one variable twice and left the other unset
- This caused the test to pass while validating the wrong condition

### Tests
- Fixed the existing incorrect test

## Commit 2 :
- Added an edge case test ensuring modifiers remain unchanged when the
  physical key changes

### CODE REVIEW

File: tests/core/input/test_shortcut.h [hunk 1]
```diff
@@ -103,7 +103,7 @@ TEST_CASE("[Shortcut] 'set_events_list' should result in the same events as the
 	CHECK(result2->get_keycode() == k2->get_keycode());
 }
 
-TEST_CASE("[Shortcut] 'matches_event' should correctly match the same event") {
+TEST_CASE("[Shortcut] 'matches_event' should correctly match the same events") {
```
Comment (AThousandShips) [inline tests/core/input/test_shortcut.h]:
```suggestion
TEST_CASE("[Shortcut] 'matches_event' should correctly match the same event") {
```
Incorrect it doesn't match multiple events

Comment (ZiedKs) [reply tests/core/input/test_shortcut.h]: This is a small misunderstanding, it is supposed to mean that it has to correctly match a singular event in all relevant case. Sorry for the mistake

File: tests/core/input/test_shortcut.h [hunk 2]
```diff
@@ -116,10 +116,10 @@ TEST_CASE("[Shortcut] 'matches_event' should correctly match the same event") {
 
 	original->set_keycode(Key::ENTER);
 	similar_but_not_equal->set_keycode(Key::ENTER);
-	similar_but_not_equal->set_keycode(Key::ESCAPE);
+	different->set_keycode(Key::ESCAPE);
 	copy = original;
```
Comment (bruvzg) [inline tests/core/input/test_shortcut.h:120]: This is just a reference to the same event, not copy, it should be:

```suggestion
	copy = original->duplicate();
```

File: tests/core/input/test_shortcut.h
Comment (bruvzg) [inline tests/core/input/test_shortcut.h:108]: Probably worth renaming it to something like `same_keycode`, `similar_but_not_equal` is not reflecting its purpose in a clear way.

Comment (bruvzg) [inline tests/core/input/test_shortcut.h:129]: This is also not doing a copy, and can be removed, just use exiting instances instead of `e_` in the testes below.

File: tests/core/input/test_input_event_key.h [hunk 3]
```diff
@@ -76,6 +76,23 @@ TEST_CASE("[InputEventKey] Key correctly stores and retrieves keycode with modif
 	CHECK(key.get_physical_keycode_with_modifiers() != Key::SPACE);
 }
 
+TEST_CASE("[InputEventKey] Modifiers persist after physical key change") {
+	InputEventKey key;
+
+	key.set_keycode(Key::ENTER);
```
Comment (bruvzg) [inline tests/core/input/test_input_event_key.h:82]: It probably should be `physical` as well, otherwise it's just changing unrelated value. Not sure what the purpose of this test, modifiers are separate from keycodes.

File: tests/core/input/test_input_event_key.h [hunk 4]
```diff
@@ -76,6 +76,23 @@ TEST_CASE("[InputEventKey] Key correctly stores and retrieves keycode with modif
 	CHECK(key.get_physical_keycode_with_modifiers() != Key::SPACE);
 }
 
+TEST_CASE("[InputEventKey] Modifiers persist after physical key change") {
+	InputEventKey key;
```
Comment (AThousandShips) [inline tests/core/input/test_input_event_key.h:80]:
```suggestion
	Ref<InputEventKey> key;
	key.instantiate();
```
And use pointer semantics, see:
* https://github.com/godotengine/godot/pull/98635
